### ISSUE TITLE
SvgDocument.Draw() Poor Text Rendering

### ISSUE THREAD

**Issue by MelanieGinger** [vvvv/SVG#253]
When I use SvgDocument.Draw(), everything is rendered ok apart from the text, the text is very grainy and poor quality. I've tried changing the PPI and increasing the size but it doesn't seem to make a difference.

I'm using version 2.2.1 and I am trying to create a bitmap image (either png or jpg) from an existing SVG file. I am opening the document and calling the Draw method, then saving the bitmap as either a jpg or a png like so:

```
var svgDocument = SvgDocument.FromSvg<SvgDocument>(svgDoc.InnerXml);
var bitmap = svgDocument.Draw();

bitmap.Save(directory, ImageFormat.Png);
```

Is there a better way to do this or a way to improve the quality or some sort of workaround for getting clear text in my images?

Thanks

![dotchart2](https://cloud.githubusercontent.com/assets/11317478/19729587/7526f286-9b8f-11e6-9475-f1cbd3680aad.png)

**Comment by FLoibl** [comment]
Having the same Problem here, Text looks far too grainy, @MelanieGinger did you find a workaround or anything?

**Comment by MelanieGinger** [comment]
@FLoibl No I've not found a fix for this yet. I'm currently going through the process of converting the svg to a Windows Metafile and trying to render the image from there. I'll let you know how it goes. 

I did have a look at those issues but a)these say the fix for this has been merged into the latest version which I already have and b)for me the text looks grainy on any background not just coloured.

**Comment by FLoibl** [comment]
@MelanieGinger I've gone back to 2.1.0 nuget in the meantime, that's the last version without the Bug imho.

**Comment by MelanieGinger** [comment]
I may well do the same.

**Comment by FLoibl** [comment]
Just out of curiosity, how would you go about converting SVG to EMF without losing Information? Running Inkscape as Process?

**Comment by MelanieGinger** [comment]
I am using the code that can be found in MetafileRenderingTest which is in SVG/Tests/Svg.UnitTests/MetafileRenderingTest.cs. I copied the DrawSvg(SvgDocument svgDoc) method which looks like this:

```
protected override Image DrawSvg(SvgDocument svgDoc)
        {
            // GDI+
            Metafile metafile;
            using (var stream = new MemoryStream())
            using (var img = new Bitmap((int)svgDoc.Width.Value, (int)svgDoc.Height.Value)) // Not necessary if you use Control.CreateGraphics().
            using (Graphics ctrlGraphics = Graphics.FromImage(img)) // Control.CreateGraphics()
            {
                IntPtr handle = ctrlGraphics.GetHdc();

                var rect = new RectangleF(0, 0, svgDoc.Width, svgDoc.Height);
                metafile = new Metafile(stream,
                    handle,
                    rect,
                    MetafileFrameUnit.Pixel,
                    EmfType.EmfPlusOnly);

                using (Graphics ig = Graphics.FromImage(metafile))
                {
                    svgDoc.Draw(ig);
                }

                ctrlGraphics.ReleaseHdc(handle);
            }

            return metafile;
        }
protected override Image DrawSvg(SvgDocument svgDoc)
        {
            // GDI+
            Metafile metafile;
            using (var stream = new MemoryStream())
            using (var img = new Bitmap((int)svgDoc.Width.Value, (int)svgDoc.Height.Value)) // Not necessary if you use Control.CreateGraphics().
            using (Graphics ctrlGraphics = Graphics.FromImage(img)) // Control.CreateGraphics()
            {
                IntPtr handle = ctrlGraphics.GetHdc();

                var rect = new RectangleF(0, 0, svgDoc.Width, svgDoc.Height);
                metafile = new Metafile(stream,
                    handle,
                    rect,
                    MetafileFrameUnit.Pixel,
                    EmfType.EmfPlusOnly);

                using (Graphics ig = Graphics.FromImage(metafile))
                {
                    svgDoc.Draw(ig);
                }

                ctrlGraphics.ReleaseHdc(handle);
            }

            return metafile;
        }
```

I've just tried using this by passing my svgDocument in and them saving the resulting image to a .png, it works but the text is still fuzzy. Note: I am still using 2.21, the next thing I am going to try is reverting back to version 2.1.0.

**Comment by MelanieGinger** [comment]
So apparently reverting back to version 2.1.0 fixes the fuzzy text issue, this is definitely a bug with version 2.2.1 (not tried 2.2.0).

**Comment by FLoibl** [comment]
2.2.0 is also broken, i tried it.

**Comment by MelanieGinger** [comment]
Cheers @FLoibl for suggesting changing the version, I've been stuck trying to find ways round using jpg/png images, unfortunately not everything supports svg images.

**Comment by tebjan** [comment]
hello, thanks for finding the right version... i might have to revert the changes on how the layers are composed. or find a better algorithm. anyone up for it? :)

**Comment by FLoibl** [comment]
@tebjan somebody mentioned something about TextCompositing in .NET in one of the related issues i mentioned above.

**Comment by tebjan** [comment]
please try with latest nuget 2.3.0, there was a problem with text rendering that might now be fixed.

**Comment by jonthysell** [comment]
I use BitmapEncoder when creating actual image files, seems to give more reliable results. See ExportImageFile in https://github.com/jonthysell/Chordious/blob/master/Chordious.WPF/Utils/ImageUtils.cs

**Comment by FLoibl** [comment]
@jonthysell Holy heck, i like that approach 👍

**Comment by radamand** [comment]
Text rendering is fixed in 2.3.0 but there is OutOfMemory Exception drawing small objects. Could you fix it in SvgVisualElements.cs or share a link to source code for 2.3.0 to fit it myself?
try
{
renderer.DrawPath(pen, path);
}
catch { }

**Comment by mrbean-bremen** [comment]
Closing - the initial issue is fixed, and problem mentioned in the last comment has a separate issue.
